src/Datadog.NativeLoader/cor_profiler.cpp

```diff
@@ -0,0 +1,874 @@
+#include "cor_profiler.h"
+
+#include "logging.h"
+
+namespace datadog
+{
+namespace nativeloader
+{
+
+    CorProfiler::CorProfiler(DynamicDispatcher* dispatcher) : ref_count_(0), dispatcher(dispatcher)
+    {
+        Debug("CorProfiler::.ctor");
+    }
+
+    CorProfiler::~CorProfiler()
+    {
+    }
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::QueryInterface(REFIID riid, void** ppvObject)
+    {
+        Debug("CorProfiler::QueryInterface");
+        if (riid == __uuidof(ICorProfilerCallback10) || riid == __uuidof(ICorProfilerCallback9) ||
+            riid == __uuidof(ICorProfilerCallback8) || riid == __uuidof(ICorProfilerCallback7) ||
+            riid == __uuidof(ICorProfilerCallback6) || riid == __uuidof(ICorProfilerCallback5) ||
+            riid == __uuidof(ICorProfilerCallback4) || riid == __uuidof(ICorProfilerCallback3) ||
+            riid == __uuidof(ICorProfilerCallback2) || riid == __uuidof(ICorProfilerCallback) || riid == IID_IUnknown)
+        {
+            *ppvObject = this;
+            this->AddRef();
+            return S_OK;
+        }
+
+        *ppvObject = nullptr;
+        return E_NOINTERFACE;
+    }
+
+    ULONG STDMETHODCALLTYPE CorProfiler::AddRef(void)
+    {
+        Debug("CorProfiler::AddRef");
+        return std::atomic_fetch_add(&this->ref_count_, 1) + 1;
+    }
+
+    ULONG STDMETHODCALLTYPE CorProfiler::Release(void)
+    {
+        Debug("CorProfiler::Release");
+        int count = std::atomic_fetch_sub(&this->ref_count_, 1) - 1;
+
+        if (count <= 0)
+        {
+            delete this;
+        }
+
+        return count;
+    }
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::Initialize(IUnknown* pICorProfilerInfoUnk)
```

Comment:
do we want to keep this naming ? (having a `p` in front of a variable name of type pointer)
To be honest, I'm not fond of it (but Greg might have another opinion on that)

Reply:
I copied the naming from the tracer, any suggestion?

Reply:
ok, let's keep it this way (tracer way), we can modify it later if we don't agree on the naming :)

Reply:
sure ;)